File: Advanced/Arrays/Logic Building/2_remove_duplicates_from_sorted_array.py
# Remove duplicates from sorted array


# Brute Force Approach
# TC = O(n log n)
# SC = O(n)
def remove_duplicates_brute(arr):
    n = len(arr)
    set_data = set()
    for i in range(n):
        set_data.add(arr[i])
    arr[:] = list(set_data)
    arr.sort()
    return len(arr)


# Optimal Approach
# TC = O(n)
# SC = O(1)
def remove_duplicates_optimal(arr):
    n = len(arr)
    if n == 0:
        return 0
    i = 0
    for j in range(1, n):
        if arr[j] != arr[i]:
            i += 1
            arr[i] = arr[j]
    return i + 1

File: Advanced/Arrays/Logic Building/2_remove_duplicates_from_sorted_array.cpp
#include <bits/stdc++.h>
using namespace std;

// --- Brute Force Approach ---
// TC = O(n log n)
// SC = O(n)
int removeDuplicatesBrute(vector<int> &arr)
{
    int n = arr.size();
    set<int> s;

    // Insert all elements into the set (automatically removes duplicates)
    for (int i = 0; i < n; i++)
    {
        s.insert(arr[i]);
    }

    // Copy unique elements back to the array
    int index = 0;
    for (int val : s)
    {
        arr[index++] = val;
    }

    // Sort to maintain sorted order
    sort(arr.begin(), arr.begin() + index);

    return index; // New length
}

// --- Optimal Approach ---
// TC = O(n)
// SC = O(1)
int removeDuplicatesOptimal(vector<int> &arr)
{
    int n = arr.size();
    if (n == 0)
        return 0;

    int i = 0;
    for (int j = 1; j < n; j++)
    {
        if (arr[j] != arr[i])
        {
            i++;
            arr[i] = arr[j];
        }
    }
    return i + 1; // New length
}

// --- Helper to Print First k Elements ---
void printArray(const vector<int> &arr, int k)
{
    for (int i = 0; i < k; i++)
        cout << arr[i] << " ";
    cout << endl;
}

// --- Main Function ---
int main()
{
    vector<int> arr1 = {1, 1, 2, 2, 3, 3, 4, 4, 5};
    vector<int> arr2 = arr1;

    cout << "Original Array: ";
    printArray(arr1, arr1.size());

    int lenBrute = removeDuplicatesBrute(arr1);
    cout << "After Brute Force: ";
    printArray(arr1, lenBrute);
    cout << "New Length: " << lenBrute << endl;

    int lenOptimal = removeDuplicatesOptimal(arr2);
    cout << "After Optimal: ";
    printArray(arr2, lenOptimal);
    cout << "New Length: " << lenOptimal << endl;

    return 0;
}
